src/tarski/syntax/terms.py

```diff
@@ -187,6 +198,167 @@ def is_syntactically_equal(self, other):
+
+    __repr__ = __str__
+
+    def __hash__(self):
+        return hash((self.symbol.symbol, tuple(x for x in self.subterms)))
+
+    def is_syntactically_equal(self, other):
+        if (self.__class__ is not other.__class__ or self.symbol != other.symbol
+                or len(self.bound_vars) != len(other.bound_vars)):
+            return False
+
+        return all(x.is_syntactically_equal(y) for x, y in zip(self.bound_vars, other.bound_vars)) \
+               and self.subterm.is_syntactically_equal(other.subterm)
+
+
+class IfThenElse(Term):
+    """
+        Combination of a formula C and two terms, t1 and t2.
+
+        This term evaluates to t1 if C is true, and t2 otherwise. t1 and t2
+        are restricted to have the same codomain.
+    """
+
+    def __init__(self, condition, subterms: Tuple[Term]):
+
+        self.condition = condition
+        if len(subterms) != 2:
+            raise err.ArityMismatch(self.symbol, subterms, msg='IfThenElse: needs two sub terms!')
+
+        self.symbol = subterms[0].language.get('ite')
+
+        # Our implementation of ite requires both branches to have equal sort
+        if subterms[0].sort != subterms[1].sort:
+            if parent(subterms[0].sort) == subterms[1].sort:
+                self._sort = subterms[1].sort
+            elif parent(subterms[1].sort) == subterms[0].sort:
+                self._sort = subterms[0].sort
+            else:
+                raise err.SyntacticError(
+                    msg='IfThenElse: both subterms need to be of the same sort! lhs: "{}"({}), rhs: "{}"({})'.format(
+                        subterms[0], subterms[0].sort, subterms[1], subterms[1].sort))
+        else:
+            self._sort = subterms[0].sort
+
+        self.subterms = tuple(subterms)
+
+        self.symbol.language.language_components_frozen = True
+
+    @property
+    def language(self):
+        return self.symbol.language
+
+    @property
+    def sort(self):
+        return self._sort
+
+    def __str__(self):
+        return '{}({})'.format(self.symbol, ', '.join([str(self.condition)] + [str(t) for t in self.subterms]))
+
+    __repr__ = __str__
+
+    def __hash__(self):
+        return hash(('ite', self.condition, tuple(x for x in self.subterms)))
+
+    def is_syntactically_equal(self, other):
+        if (self.__class__ is not other.__class__ or self.symbol != other.symbol \
+            or len(self.subterms) != len(other.subterms)) \
+                or (not self.condition.is_syntactically_equal(other.condition)):
+            return False
+
+        # Else we just need to recursively check if all subterms are syntactically equal
+        return all(x.is_syntactically_equal(y) for x, y in zip(self.subterms, other.subterms))
+
+
+def ite(c, t1, t2):
+    return IfThenElse(c, (t1, t2))
+
+
+class Matrix(Term):
+
+    def __init__(self, arraylike, sort: Sort):
+        self.matrix = np.matrix(arraylike, dtype=np.dtype(object))
```

Comment:
This line gives me endless warnings of the form:

    PendingDeprecationWarning: the matrix subclass is not the recommended way to represent matrices or deal with linear algebra (see https://docs.scipy.org/doc/numpy/user/numpy-for-matlab-users.html). Please adjust your code to use regular ndarray

Should we switch to ndarrays? (If we are using this class, of course)

Reply:
That's a warning that appears on Python 3.7, and I wasn't aware of it until I eventually got the latest version of Ubuntu. Here's a quite indepth explanation of the change

https://stackoverflow.com/questions/53254738/deprecation-status-of-the-numpy-matrix-class

Addressing this warning has been quite complicated - see commit b3464b943ee472c7b6a8bdf77e40080b45f4d94f - as ```numpy.ndarray``` does not support multiplication of matrices holding arbitrary objects (while the older ```numpy.matrix``` type did). So I have had to implement a simple routine that allows to multiply and add arbitrary Python objects.